database/sqlite/sqlite_functions.c

```diff
@@ -107,13 +114,39 @@ static void add_stmt_to_list(sqlite3_stmt *res)
 
     if (unlikely(idx == MAX_OPEN_STATEMENTS))
         return;
-    statements[idx++] = res;
 }
 
-int prepare_statement(sqlite3 *database, char *query, sqlite3_stmt **statement) {
+static void release_statement(void *statement)
+{
+    int rc;
+#ifdef NETDATA_INTERNAL_CHECKS
+    info("Thread %d: Cleaning prepared statement on %p", gettid(), statement);
+#endif
+    if (unlikely(rc = sqlite3_finalize((sqlite3_stmt *) statement) != SQLITE_OK))
+        error_report("Failed to finalize statement, rc = %d", rc);
+}
+
+int prepare_statement(sqlite3 *database, char *query, sqlite3_stmt **statement)
+{
+    static __thread uint32_t keys_used = 0;
+
+    pthread_key_t *key = NULL;
+    int ret = 1;
+
+    if (likely(keys_used < MAX_PREPARED_STATEMENTS))
+        key = &key_pool[keys_used++];
+
     int rc = sqlite3_prepare_v2(database, query, -1, statement, 0);
-    if (likely(rc == SQLITE_OK))
-        add_stmt_to_list(*statement);
+    if (likely(rc == SQLITE_OK)) {
+        if (likely(key)) {
+            ret = pthread_setspecific(*key, *statement);
+#ifdef NETDATA_INTERNAL_CHECKS
+            info("Thread %d: Using key %u on statement %p", gettid(), keys_used, *statement);
+#endif
+        }
+        if (ret)
+            add_stmt_to_list(*statement);
+    }
```

Comment:
Stelio hi! I'm not sure, so we support up to 32 prepared statements, and after that the rest will be prepared but not added to the list, and so won't get cleaned up? I assume 32 will cover our current sql queries, but we need to adjust the number if we add more, etc ?

Reply:
yes, so it is 32 per thread -- right now we have 17 cases total of calls to `prepare_statement`